### PR TITLE
Upgrade to lucene-7.2.0-snapshot-8c94404.

### PR DESCRIPTION
**Pull request by jpountz**
The main highlight of this new snapshot is that it introduces the opportunity
for queries to opt out of caching. In case a query opts out of caching, not only
will it never be cached, but also no compound query that wraps it will be
cached.

### CODE REVIEW

File: core/src/main/java/org/apache/lucene/queries/SearchAfterSortedDocQuery.java [hunk 1]
```diff
@@ -90,6 +88,14 @@ public Scorer scorer(LeafReaderContext context) throws IOException {
                 final DocIdSetIterator disi = new MinDocQuery.MinDocIterator(firstDoc, maxDoc);
                 return new ConstantScoreScorer(this, score(), disi);
             }
+
+            @Override
+            public boolean isCacheable(LeafReaderContext ctx) {
+                // If the sort order includes _doc, then the matches in a segment
+                // may depend on other segments, which makes this query a bad
+                // candidate for caching
```
Comment (jimczi) [inline core/src/main/java/org/apache/lucene/queries/SearchAfterSortedDocQuery.java]: ++

File: core/src/main/java/org/elasticsearch/index/query/ScriptQueryBuilder.java [hunk 2]
```diff
@@ -196,6 +196,11 @@ public float matchCost() {
                     };
                     return new ConstantScoreScorer(this, score(), twoPhase);
                 }
+
+                @Override
+                public boolean isCacheable(LeafReaderContext ctx) {
+                    return true;
```
Comment (jimczi) [inline core/src/main/java/org/elasticsearch/index/query/ScriptQueryBuilder.java]: Should we return false and change the implementation of equals and hashCode if we still don't want to cache script query ?

File: core/src/main/java/org/elasticsearch/index/query/TermsSetQueryBuilder.java [hunk 3]
```diff
@@ -290,7 +291,6 @@ public boolean needsScores() {
 
         @Override
         public int hashCode() {
-            // CoveringQuery with this field value source cannot be cachable
```
Comment (jimczi) [inline core/src/main/java/org/elasticsearch/index/query/TermsSetQueryBuilder.java]: Same here, probably not important but we can use the script for hashCode and equals ?

**Review by jimczi — COMMENTED** [review]
Thanks @jpountz 
I left some comments regarding some additional cleanup that we can do now that opting out from the cache is explicit.

**Comment by jpountz** [comment]
@elasticmachine please test it

**Review by jimczi — APPROVED** [review]
LGTM

**Pull request merged**